flight/Modules/FlightStats/flightstatsmodule.c

```diff
@@ -0,0 +1,257 @@
+
+// Private variables
+static bool module_enabled;
+static struct pios_thread *flightStatsTaskHandle;
+static FlightStatsSettingsData settings;
+static PositionActualData lastPositionActual;
+static float initial_consumed_energy;
+
+// Private functions
+static void flightStatsTask(void *parameters);
+static void settingsUpdatedCb(UAVObjEvent * ev);
+static bool isArmed();
+static void collectStats(FlightStatsData *stats);
+
+// Local variables
+
+// External variables
+
+/**
+ * Initialize the FlightStats module
+ * \return -1 if initialisation failed
+ * \return 0 on success
+ */
+int32_t FlightStatsModuleInitialize(void)
+{
+#ifdef MODULE_FLIGHTSTATS_BUILTIN
+	module_enabled = true;
+#else
+	uint8_t module_state[MODULESETTINGS_ADMINSTATE_NUMELEM];
+	ModuleSettingsAdminStateGet(module_state);
+	if (module_state[MODULESETTINGS_ADMINSTATE_FLIGHTSTATS] == MODULESETTINGS_ADMINSTATE_ENABLED) {
+		module_enabled = true;
+	} else {
+		module_enabled = false;
+	}
+#endif
+
+	if (!module_enabled)
+		return -1;
+
+	FlightStatsInitialize();
+	FlightStatsSettingsInitialize();
+
+	// Get settings and connect callback
+	FlightStatsSettingsGet(&settings);
+	FlightStatsSettingsConnectCallback(settingsUpdatedCb);
+
+	return 0;
+}
+
+/**
+ * Start the FlightStats module
+ * \return -1 if initialization failed
+ * \return 0 on success
+ */
+int32_t FlightStatModuleStart(void)
+{
+	//Check if module is enabled or not
+	if (module_enabled == false) {
+		return -1;
+	}
+
+	// Start flight stats task
+	flightStatsTaskHandle = PIOS_Thread_Create(flightStatsTask, "FlightStats", STACK_SIZE_BYTES, NULL, TASK_PRIORITY);
+
+	TaskMonitorAdd(TASKINFO_RUNNING_FLIGHTSTATS, flightStatsTaskHandle);
+	
+	return 0;
+}
+
+MODULE_INITCALL(FlightStatsModuleInitialize, FlightStatModuleStart);
+
+static void flightStatsTask(void *parameters)
+{
+	bool first_init = true;
+	FlightStatsData flightStatsData;
+
+	flightStatsData.State = FLIGHTSTATS_STATE_IDLE;
+
+	// Loop forever
+	while (1) {
+		// Update stats at about 10Hz
+		PIOS_Thread_Sleep(100);
+		switch (flightStatsData.State) {
+			case FLIGHTSTATS_STATE_IDLE:
+				if (isArmed()) {
+					switch (settings.StatsBehavior) {
+						case FLIGHTSTATSSETTINGS_STATSBEHAVIOR_RESETONBOOT:
+							if (first_init) {
+								flightStatsData.State = FLIGHTSTATS_STATE_RESET;
+							}
+							else {
+								flightStatsData.State = FLIGHTSTATS_STATE_INITIALIZING;
+							}
+							break;
+						case FLIGHTSTATSSETTINGS_STATSBEHAVIOR_RESETONARM:
+							flightStatsData.State = FLIGHTSTATS_STATE_RESET;
+							break;
+					}
+				}
+				break;
+			case FLIGHTSTATS_STATE_RESET:
+				memset((void*)&flightStatsData, 0, sizeof(flightStatsData));
+				flightStatsData.State = FLIGHTSTATS_STATE_INITIALIZING;
+				break;
+			case FLIGHTSTATS_STATE_INITIALIZING:
+				PositionActualGet(&lastPositionActual);
+				if (first_init || settings.StatsBehavior == FLIGHTSTATSSETTINGS_STATSBEHAVIOR_RESETONARM) {
+					if (FlightBatteryStateHandle()) {
```

Comment:
Not critical, but this logic took me a few minutes to grok and I think can be simplified.
1. create a nice little method setInitialBattery() that does what this block of code does. Call it before entering the while loop.
2. make _STATE_RESET explicitly only reset the fields it wants to (basically everything but InitialBatteryVoltage
3. get rid of first_init variable, and in conditional of L137 go straight to _COLLECTING state (because essentially already initialized). the rest of the code can go.
4. delete the _INITIALIZING state, and call setInitialBattery() from _RESET

IMO that would be a fair bit more readable (except more lines at L152 to set the fields individually, but that's more typical of how we normally do it). Not a blocker though.